openchain/consensus/obcpbft/obc-executor.go

```diff
@@ -0,0 +1,472 @@
+	Validate   bool // Whether to validate the execution before committing
+}
+
+type syncTarget struct {
+	seqNo     uint64
+	peerIDs   []*pb.PeerID
+	blockInfo *BlockInfo
+}
+
+type transaction struct {
+	seqNo    uint64
+	txs      []*pb.Transaction // nil has a special meaning, that this represents some missing/discarded requests and should possibly state transfer
+	execInfo *ExecutionInfo
+}
+
+type obcExecutor struct {
+	executionQueue chan *transaction
+	completeSync   chan *syncTarget
+	syncTargets    chan *syncTarget
+	threadIdle     chan struct{}
+	threadExit     chan struct{}
+	lastExec       uint64
+	id             *pb.PeerID
+	orderer        Orderer
+	sts            *statetransfer.StateTransferState
+	executorStack  statetransfer.PartialStack
+}
+
+func NewOBCExecutor(config *viper.Viper, orderer Orderer, stack statetransfer.PartialStack) (obcex *obcExecutor) {
+	var err error
+	obcex = &obcExecutor{}
+	queueSize := config.GetInt("executor.queuesize")
+	if queueSize <= 0 {
+		panic("Queue size must be positive")
+	}
+
+	obcex.executionQueue = make(chan *transaction, queueSize)
+	obcex.syncTargets = make(chan *syncTarget)
+	obcex.completeSync = make(chan *syncTarget)
+	obcex.threadIdle = make(chan struct{})
+	obcex.threadExit = make(chan struct{})
+	obcex.orderer = orderer
+	obcex.executorStack = stack
+	obcex.id, _, err = stack.GetNetworkHandles()
+	if nil != err {
+		logger.Error("Could not resolve our own PeerID, assigning dummy")
+		obcex.id = &pb.PeerID{"Dummy"}
+	}
+
+	logger.Info("Executor for %v using queue size of %d", obcex.id, queueSize)
+
+	obcex.sts = statetransfer.NewStateTransferState(config, stack)
+
+	listener := struct{ statetransfer.ProtoListener }{}
+	listener.CompletedImpl = obcex.stateTransferCompleted
+	obcex.sts.RegisterListener(&listener)
+
+	go obcex.queueThread()
+	return
+}
+
+func (obcex *obcExecutor) Stop() {
+	logger.Debug("%v stopping executor", obcex.id)
+	select {
+	case <-obcex.threadExit:
+	default:
+		close(obcex.threadExit)
+	}
+	obcex.sts.Stop()
+}
+
+// Informs the other end of the syncTargets if someone is listening
+func (obcex *obcExecutor) ValidState(seqNo uint64, id []byte, peerIDs []*pb.PeerID) {
+	blockInfo := &BlockInfo{}
+	if err := proto.Unmarshal(id, blockInfo); nil != err {
+		logger.Error("%v error unmarshalling id to blockInfo: %s", obcex.id, err)
+		return
+	}
+	select {
+	case obcex.syncTargets <- &syncTarget{
+		seqNo:     seqNo,
+		peerIDs:   peerIDs,
+		blockInfo: blockInfo,
+	}:
+		logger.Debug("%v sent sync target", obcex.id)
+	default:
+	}
+}
+
+// Enqueues a request for execution if there is room
+func (obcex *obcExecutor) Execute(seqNo uint64, txs []*pb.Transaction, execInfo *ExecutionInfo) {
+	request := &transaction{
+		seqNo:    seqNo,
+		txs:      txs,
+		execInfo: execInfo,
+	}
+	select {
+	case obcex.executionQueue <- request:
+		logger.Debug("%v queued request for sequence number %d", obcex.id, seqNo)
+	default:
```

Comment:
The executionQueue is a buffered channel.  Its size is configurable, but for pbft, it should be at least the size of the pbft log.  Because there can not be more outstanding executions than the the log size (because a checkpoint is required to move the watermarks, and a checkpoint is generated from the execution result), it should never be possible for pbft to overflow this channel.

However, in the interest of generality, and to recover from any unforeseen errors, the write to the channel is done via a select in a non-blocking way.  If there is no room on the channel, then the channel is emptied, and a flag (a special request with nil txs)  is set in the channel indicating that there was a gap and that state transfer should be performed.